src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -274,10 +275,28 @@ async def mypy_typecheck_partition(
 
                             {mkdir.path} -p {run_cache_dir}/{py_version} > /dev/null 2>&1 || true
                             {cp.path} {mypy_cache_dir}/{py_version}/cache.db {run_cache_dir}/{py_version}/cache.db > /dev/null 2>&1 || true
+
                             {' '.join((shell_quote(arg) for arg in argv))}
                             EXIT_CODE=$?
+
+                            get_device_id() {{
+                                TEST_PATH=$({readlink.path} -f "$1")
+                                stat -c "%d" "$TEST_PATH" 2>/dev/null || stat -f "%d" "$TEST_PATH" 2>/dev/null
+                            }}
                             {mkdir.path} -p {mypy_cache_dir}/{py_version} > /dev/null 2>&1 || true
-                            {mv.path} {run_cache_dir}/{py_version}/cache.db {mypy_cache_dir}/{py_version}/cache.db > /dev/null 2>&1 || true
+                            SANDBOX_DEVID=$(get_device_id "{run_cache_dir}")
+                            NAMED_CACHE_DEVID=$(get_device_id "{mypy_cache_dir}")
+
+                            if [ "$SANDBOX_DEVID" = "$NAMED_CACHE_DEVID" ]; then
+                                # Same device, just `mv`
+                                {mv.path} {run_cache_dir}/{py_version}/cache.db {mypy_cache_dir}/{py_version}/cache.db > /dev/null 2>&1 || true
```

Comment:
Are we always certain that the mv/cp target doesn't exist? (I know this question was pertinent even before this change, but just checking)

Reply:
No, it most certainly can exist before the mv

Reply:
So why don't we need `mv -f` ?

Reply:
No clue 🙃

I assume MV determines there's no stdin stream and doesn't prompt.

Reply:
My understanding is that mv doesn't prompt in most cases, including if the target exists.

> An even more precise definition of how mv operates is given in the [POSIX standard](http://pubs.opengroup.org/onlinepubs/009604599/utilities/mv.html), which adds that -f only overrides -i if it occurs later in the command line.
>
> So the default behavior is a bit different from -f. The default is to ask for confirmation only when the target isn't writable. (This behavior goes back at least as far as [V4](http://minnie.tuhs.org/cgi-bin/utree.pl?file=V4/man/man1/mv.1), where mv didn't take any options.) If the -i option is given, mv will additionally ask for confirmation whenever the target exists. The -f option will inhibit asking in both of those cases (if it occurs after any -i).

https://unix.stackexchange.com/a/236676

Reply:
Also this comment from that answer:

> It's worth noting that many implementations only ask for confirmation if they appear to be in an interactive terminal (isatty(0)).